mcs/class/System/Mono.UnityTls/UnityTlsContext.cs

```diff
@@ -0,0 +1,369 @@
+		}
+
+		public override void Shutdown ()
+		{
+			// Destroy native UnityTls objects
+			UnityTls.NativeInterface.unitytls_tlsctx_free (m_TlsContext);
+			m_TlsContext = null;
+
+			m_HasContext = false;
+		}
+
+		protected override void Dispose (bool disposing)
+		{
+			try {
+				if (disposing)
+				{
+					Shutdown();
+
+					// reset states
+					m_LocalClientCertificate = null;
+					m_RemoteCertificate = null;
+					m_Connectioninfo = null;
+					m_IsAuthenticated = false;
+					m_HasContext = false;
+				}
+
+				m_handle.Free();
+
+			} finally {
+				base.Dispose (disposing);
+			}
+		}
+
+		public override void StartHandshake ()
+		{
+			// TODO: Client->Server authentification is not supported by UnityTls as of writing
+			if (IsServer && AskForClientCertificate) {
+				throw new NotImplementedException ("No support for server-sided client certificate check yet.");
+			}
+
+			if (Settings != null && Settings.EnabledCiphers != null) {
+				var ciphers = new UnityTls.unitytls_ciphersuite [Settings.EnabledCiphers.Length];
+				for (int i = 0; i < ciphers.Length; i++)
+					ciphers [i] = (UnityTls.unitytls_ciphersuite)Settings.EnabledCiphers [i];
+				
+				var errorState = UnityTls.NativeInterface.unitytls_errorstate_create ();
+				fixed (UnityTls.unitytls_ciphersuite* ciphersPtr = ciphers)
+					UnityTls.NativeInterface.unitytls_tlsctx_set_supported_ciphersuites (m_TlsContext, ciphersPtr, ciphers.Length, &errorState);
+				Unity.Debug.CheckAndThrow (errorState, "Failed to set list of supported ciphers", AlertDescription.HandshakeFailure);
+			}
+		}
+
+		public override bool ProcessHandshake ()
+		{
+			var errorState = UnityTls.NativeInterface.unitytls_errorstate_create ();
+			var result = UnityTls.NativeInterface.unitytls_tlsctx_process_handshake (m_TlsContext, &errorState);
+			if (errorState.code == UnityTls.unitytls_error_code.UNITYTLS_USER_WOULD_BLOCK)
+				return false;
+
+			// Not done is not an error if we are server and don't ask for ClientCertificate
+			if (result == UnityTls.unitytls_x509verify_result.UNITYTLS_X509VERIFY_NOT_DONE && IsServer && !AskForClientCertificate)
+				Unity.Debug.CheckAndThrow (errorState, "Handshake failed", AlertDescription.HandshakeFailure);
+			else
+				Unity.Debug.CheckAndThrow (errorState, result, "Handshake failed", AlertDescription.HandshakeFailure);
+
+			// .Net implementation gives the server a verification callback (with null cert) even if AskForClientCertificate is false.
+			// We stick to this behavior here.
+			if (IsServer && !AskForClientCertificate) {
+				if (!ValidateCertificate (null, null))
+					throw new TlsException (AlertDescription.HandshakeFailure, "Verification failure during handshake");
+			}
+
+			return true;
+		}
+
+		public override void FinishHandshake ()
+		{
+			// Query some data. Ignore errors on the way since failure is not crucial.
+			var cipherSuite = UnityTls.NativeInterface.unitytls_tlsctx_get_ciphersuite(m_TlsContext, null);
```

Comment:
passing null for the errorstate is soon no longer supported. need to change this

Reply:
[fixed]